### PR TITLE
prov/rxm: Optimize CQ + bug-fixing

### PR DESCRIPTION
**Pull request by dmitrygx**
This PR does the following things:
- optimizes CQ by implementing multi and single version of progress for each RxM EP CQ.
- fixes out-of-memory kill by Linux OOM killer in the case when user does a large number of `fi_tinject`s with the size > than supported by core provider.  It leads that RxM providers allocates a large number of TX requests (core providers completes them internally, but RxM doesn't read them, becasue cokre provider reports `FI_SUCCESS` for `fi_send`) w/o reading completion for a long time. Solution: if TX buffer pool doesn't have enough space - progress CQ and if there is still no space shrink TX buffer pool.

### CODE REVIEW

File: prov/rxm/src/rxm_rma.c [hunk 1]
```diff
@@ -459,7 +459,7 @@ static ssize_t rxm_ep_inject_writedata(struct fid_ep *ep_fid, const void *buf,
 			       "fi_inject_writedata for MSG provider failed with ret - %"
 			       PRId64"\n", ret);
 			if (OFI_LIKELY(ret == -FI_EAGAIN))
-				rxm_ep_progress(&rxm_ep->util_ep);
+				rxm_ep_progress_multi(&rxm_ep->util_ep);
```
Comment (a-ilango) [inline prov/rxm/src/rxm_rma.c]: Is this patch a revert of 053392629e3cd6f18281fe3896a0551b0fe136ec?

File: prov/rxm/src/rxm_cq.c [hunk 2]
```diff
@@ -999,12 +1000,16 @@ static ssize_t rxm_cq_handle_comp(struct rxm_ep *rxm_ep,
...
 		tx_eager_buf = comp->op_context;
 		assert(comp->flags & FI_SEND);
 		ret = rxm_finish_eager_send(rxm_ep, tx_eager_buf);
 		rxm_tx_buf_release(rxm_ep, RXM_BUF_POOL_TX, tx_eager_buf);
 		return ret;
+	case RXM_INJECT_TX:
+		tx_base_buf = comp->op_context;
+		assert(comp->flags & FI_SEND);
+		rxm_tx_buf_release(rxm_ep, RXM_BUF_POOL_TX_INJECT, tx_base_buf);
+		return 0;
```
[the remark this answers is not in the document]
Comment (shefty) [reply prov/rxm/src/rxm_cq.c]: Is this needed for 1.7.x?

File: prov/rxm/src/rxm_cq.c [hunk 3]
```diff
@@ -1267,60 +1267,120 @@ int rxm_ep_prepost_buf(struct rxm_ep *rxm_ep, struct fid_ep *msg_ep)
...
 	rxm_ep_progress_unsafe(util_ep);
 	ofi_ep_lock_release(util_ep);
 }
 
+void rxm_ep_progress_multi(struct util_ep *util_ep)
+{
+	ofi_ep_lock_acquire(util_ep);
+	rxm_ep_progress_multi_unsafe(util_ep);
+	ofi_ep_lock_release(util_ep);
+}
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: Please find a way to structure these 4 functions without duplicating so much code.  We're just asking to introduce the same bug multiple times.

File: prov/util/src/util_buf.c [hunk 4]
```diff
@@ -49,6 +49,17 @@ int util_buf_grow(struct util_buf_pool *pool)
 	ssize_t hp_size;
 	struct util_buf_footer *buf_ftr;
 
+	if ((pool->num_allocated > 0) && pool->attr.unavail_hndlr) {
```
Comment (a-ilango) [inline prov/util/src/util_buf.c]: Did you mean `pool->attr.max_cnt && pool->num_allocated >= pool->attr.max_cnt && pool->attr.unavail_hndlr` like the check below?

File: prov/rxm/src/rxm_cq.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (a-ilango) [reply prov/rxm/src/rxm_cq.c]: Yes, this would be needed for 1.7.x.

File: include/ofi_mem.h [hunk 5]
```diff
@@ -273,6 +274,7 @@ struct util_buf_attr {
 	size_t 				chunk_cnt;
 	util_buf_region_alloc_hndlr 	alloc_hndlr;
 	util_buf_region_free_hndlr 	free_hndlr;
+	util_buf_region_unavail_hndlr	unavail_hndlr;
```
Comment (shefty) [inline include/ofi_mem.h]: We already have a alloc_hndlr.  This looks like a pre_alloc_handler.  In many cases, the user could perform the necessary actions as part of alloc_handler.  The pool would grow first, but I'm not sure that really matters.

**Comment by shefty** [comment]
Old PR with several conflicts from a lost developer.  :(  Closing.

**Pull request closed without merging**